src/Symfony/Component/JsonPath/JsonCrawler.php

```diff
@@ -39,16 +39,23 @@ final class JsonCrawler implements JsonCrawlerInterface
     ];
 
     /**
-     * @param resource|string $raw
+     * @param resource|string|array $data
      */
-    public function __construct(
-        private readonly mixed $raw,
-    ) {
-        if (!\is_string($raw) && !\is_resource($raw)) {
-            throw new InvalidArgumentException(\sprintf('Expected string or resource, got "%s".', get_debug_type($raw)));
+    public function __construct(private mixed $data = [])
+    {
+        if (!\is_string($data) && !\is_resource($data) && !\is_array($data)) {
+            throw new \TypeError(\sprintf('Argument #1 ($data) must be of type string, array or resource, %s given.', get_debug_type($data)));
         }
     }
 
+    /**
+     * @param resource|string|array $data
+     */
+    public function fromJson(mixed $data): CrawlerInterface
```

Comment:
What about making the method static? And I would question if we really need an interface for it? And if so, if it's name wouldn't better be `JsonCrawlerFactoryInterface`.

Reply:
Making the method static wouldn't allow us to plug properly with DI and allow further steps like [this one](https://github.com/symfony/symfony/pull/60083#discussion_r2055527025)

Reply:
In your last snippet, you call `withCustomFunctions()` yourself, but the goal is to have everything wired correctly so custom functions are automatically registered and you "never" call the function yourself if you use FWB. I'm not sure we'd be able to achieve this with static factories (or any static behavior being used)?

Reply:
You might be able to combine static factories and [service configurators](https://symfony.com/doc/current/service_container/configurators.html) maybe?

Reply:
Hell no, not configurators.
A static factory would miss the point of configuring transformer functions.

But all this makes me think we're thinking out of the blue. Let's not do anything until we actually do have such transformers.

If we have some lessons to take from this PR to prepare the next steps, let's keep that, but let's remove the autowiring/DI part to the moment where we'll really need it.